c7n/policy.py

```diff
@@ -1262,7 +1262,12 @@ def get_variables(self, variables=None):
             'op': '{op}',
             'action_date': '{action_date}',
             # tag action pyformat-date handling
-            'now': utils.FormatDate(datetime.utcnow()),
+            # defer expansion until runtime for serverless modes
+            'now': (
+                utils.DeferredFormatString('now')
```

Comment:
its not clear how this would resolve the original issue wrt to json dump type error, we need a test better simulating e2e for that. I suspect the issue is actually separate, and in notify's message serialization.

Reply:
fwiw, I tried modifying one of the test_notify but it successfully ran so its not clear on the original issue without a traceback I think.

```diff
diff --git a/tests/test_notify.py b/tests/test_notify.py
index 9e93ee490..2be90861a 100644
--- a/tests/test_notify.py
+++ b/tests/test_notify.py
@@ -44,6 +44,7 @@ class NotifyTest(BaseTest):
                     {
                         "type": "notify",
                         "to": ["to@example.com"],
+                        "date": "{now}",
                         "to_from": {
                             "url": url,
                             "format": "json",
```

Reply:
Fair, I'll look to get a test together. It looked to me like it was interpolating a FormatDate and then dying when mu tried to serialize the policy to config.json while building the function package.

Reply:
Okie added a test - it was dying dumping policy details to JSON as part of `get_archive()` [here](https://github.com/cloud-custodian/cloud-custodian/blob/ef53a606f855b4d811caa3a26f92f6bd624a61cc/c7n/mu.py#L962).

When I added the test I realized that I was initially focused on not losing format specifiers, so didn't have a test _without_ them. The test without them failed because `VarFormat` was passing the DeferredFormatString object through directly.

If `{now}` is the _only_ special case where we want to avoid interpolation at provision time, maybe it would be cleaner to explicitly handle that case inside `VarFormat` and not have the `DeferredFormatString` class in the first place :shrug: . I'm not sure what's clearer to other folks and future us.